### PR TITLE
feat(trustpayments): implement ApplePay wallet support

### PR DESCRIPTION
**Pull request by 10X-GRACE**
## Summary

Adds Apple Pay (decrypted) wallet support to the **trustpayments** connector. Apple Pay is mapped onto Trust Payments' wallet-sourced AUTH request (\`walletsource=APPLEPAY\`, \`tokentype=APPLEPAY\`, DPAN passthrough, TAVV cryptogram, ECI). Trust Payments has no native encrypted Apple Pay endpoint, so the decrypted-passthrough pattern is used.

## Changes

- \`crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs\`: new \`PaymentMethodData::Wallet(WalletData::ApplePay(_))\` arm in the Authorize request transformer; constructs \`TrustpaymentsApplePayData { pan, expirydate (MM/YYYY), tavv, eci, tokenisedpayment, tokentype, walletsource, walletdisplayname }\`. Defaults ECI to \`07\` (Apple Pay non-3DS) when the wallet payload omits it. Existing card + Google Pay paths untouched.
- Added \`walletdisplayname\` field to \`TrustpaymentsApplePayData\` — TrustPayments requires this field in the AUTH request for Apple Pay; populated from \`apple_pay_data.payment_method.display_name\` (e.g. \`"Visa 4242"\`).
- Cargo \`fmt\` applied to the same file.

### Layer 1: Direct gRPC (UCS → TrustPayments)

Multiple gRPC scenarios run directly against UCS (\`localhost:8000\`) with decrypted Apple Pay data, all returned **CHARGED**:

| Card Network | Currency | Transaction Ref | walletdisplayname |
|---|---|---|---|
| Visa | GBP | `57-9-4851895` | Visa 1111 |
| Mastercard | USD | verified CHARGED | MC 4444 |
| Amex | EUR | verified CHARGED | Amex 0005 |

<details>
<summary>grpcurl command (sanitized)</summary>

```bash
grpcurl -plaintext \
  -H "x-connector: <REDACTED>" \
  -H "x-auth: <REDACTED>" \
  -H "x-api-key: <REDACTED>" \
  -H "x-api-secret: <REDACTED>" \
  -H "x-key1: <REDACTED>" \
  -H "x-merchant-id: <REDACTED>" \
  -H "x-reference-id: <REDACTED>" \
  -d '{
    "merchant_transaction_id": "ref_pr_tp_applepay_03",
    "amount": {"minor_amount": 1000, "currency": "GBP"},
    "capture_method": "AUTOMATIC",
    "auth_type": "NO_THREE_DS",
    "payment_method": {
      "apple_pay": {
        "payment_data": {
          "decrypted_data": {
            "application_primary_account_number": {"value": "<REDACTED_DPAN>"},
            "application_expiration_month": {"value": "03"},
            "application_expiration_year": {"value": "2030"},
            "payment_data": {
              "online_payment_cryptogram": {"value": "<REDACTED_CRYPTOGRAM>"},
              "eci_indicator": "07"
            }
          }
        },
        "payment_method": {"display_name": "Visa 1111", "network": "visa", "type": "debit"},
        "transaction_identifier": "tx_pr_tp_applepay_03"
      }
    },
    "address": {"billing_address": {"country_alpha2_code": "GB", "first_name": {"value": "John"}, "last_name": {"value": "Doe"}}},
    "return_url": "https://example.com/return",
    "webhook_url": "https://example.com/webhook",
    "order_category": "pay",
    "enrolled_for_3ds": false,
    "request_incremental_authorization": false
  }' \
  localhost:8000 types.PaymentService/Authorize
```

</details>

<details>
<summary>grpcurl response (sanitized)</summary>

```json
{
  "merchantTransactionId": "W57-u2fQYvHj",
  "connectorTransactionId": "57-9-4851895",
  "status": "CHARGED",
  "statusCode": 200,
  "rawConnectorResponse": {
    "value": "{\n    \"requestreference\": \"W57-u2fQYvHj\",\n    \"response\": [\n        {\n            \"accounttypedescription\": \"ECOM\",\n            \"acquirerresponsecode\": \"00\",\n            \"authcode\": \"TEST48\",\n            \"baseamount\": \"1000\",\n            \"credentialsonfile\": \"1\",\n            \"currencyiso3a\": \"GBP\",\n            \"eci\": \"07\",\n            \"errorcode\": \"0\",\n            \"errormessage\": \"Ok\",\n            \"maskedpan\": \"411111######1111\",\n            \"orderreference\": \"ref_pr_tp_applepay_03\",\n            \"paymenttypedescription\": \"VISA\",\n            \"requesttypedescription\": \"AUTH\",\n            \"settlestatus\": \"0\",\n            \"tavv\": \"<REDACTED_CRYPTOGRAM>\",\n            \"tokenisedpayment\": \"1\",\n            \"tokentype\": \"APPLEPAY\",\n            \"transactionreference\": \"57-9-4851895\",\n            \"walletsource\": \"APPLEPAY\",\n            \"walletdisplayname\": \"Visa 1111\"\n        }\n    ],\n    \"version\": \"1.00\"\n}"
  }
}
```

</details>

### Layer 2: E2E cURL (HS Router → UCS → TrustPayments)

Full end-to-end test via HS Router (\`localhost:8080\`) with a real Apple Pay token (encrypted with PPC cert, decrypted by HS router before being forwarded to UCS as DPAN):

- HS Router decrypted the Apple Pay token using the PPC certificate
- Forwarded DPAN + TAVV + ECI to UCS as \`ApplePayDecrypt\`
- UCS constructed the TrustPayments AUTH request with \`walletsource=APPLEPAY\`, \`walletdisplayname="Visa 4242"\`
- TrustPayments returned **CHARGED**

Key response fields confirmed:
- \`status\`: \`succeeded\`
- \`requesttypedescription\`: \`AUTH\`
- \`walletsource\`: \`APPLEPAY\`
- \`walletdisplayname\`: \`"Visa 4242"\`
- \`errorcode\`: \`0\` ("Ok")

> **Note on Apple Pay E2E testing:** Apple Pay tokens cannot be generated in a browser on localhost — this requires HTTPS, domain verification, and Safari on Apple hardware with Touch ID/Face ID. Testing is done in two layers: (1) direct gRPC to UCS with pre-decrypted DPAN data, and (2) E2E via HS Router using a real encrypted token (encrypted with the PPC cert configured in the router) sent via cURL.

## Test plan

- [x] \`cargo build --package connector-integration\` — clean
- [x] \`cargo clippy --package connector-integration -- -D warnings\` — no trustpayments warnings (1 pre-existing hipay warning unrelated to this PR)
- [x] \`cargo +nightly fmt\` — only trustpayments formatting changes retained
- [x] gRPC Apple Pay Authorize (Visa/GBP, Mastercard/USD, Amex/EUR) against Trust Payments sandbox — all REAL_PASS, CHARGED
- [x] E2E cURL via HS Router → UCS → TrustPayments — REAL_PASS, \`status: succeeded\`, \`walletdisplayname: "Visa 4242"\`

### CODE REVIEW

File: crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs [hunk 1]
```diff
@@ -169,6 +169,7 @@ pub struct TrustpaymentsAuthRequest {
 pub enum TrustpaymentsPaymentMethod {
     Card(TrustpaymentsCardData),
     GooglePay(Box<TrustpaymentsGooglePayData>),
+    ApplePay(Box<TrustpaymentsApplePayData>),
```
Comment (maverox) [inline crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs:172]: 🟡 **S2 — trustpayments ApplePay: verify pre-decrypted data requirement and DPAN handling**

Same pattern as other ApplePay PRs. Verify: (1) the connector requires pre-decrypted Apple Pay data, (2) the DPAN is correctly mapped to the connector's card number field, (3) expiry format matches connector requirements. Scope drift also applies.

**Comment by Shubhodip900** [comment]
Code review
Found 4 issues:

1. HIGH - Magic string for ECI default value

In the Apple Pay payment method handler, the ECI default value "07" is hardcoded inline. This should be defined as a constant for maintainability and to prevent typos.

crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs

Lines 380-382

    // For non-cryptogram flows, Trust Payments requires an ECI value;
    // default to "07" (Apple Pay non-3DS) if absent.
    eci: Some(eci.unwrap_or_else(|| "07".to_string())),

Suggested fix:

    const APPLE_PAY_NON_3DS_ECI: &str = "07";
    
    // Then use:
    eci: Some(eci.unwrap_or_else(|| APPLE_PAY_NON_3DS_ECI.to_string())),

---

2. MEDIUM - Duplicate string literals "APPLEPAY"

The literal "APPLEPAY" appears multiple times in the Apple Pay handler. This should use a constant to ensure consistency.

crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs

Lines 389, 394, 397

    tokentype: if is_cryptogram_3ds {
        Some("APPLEPAY".to_string())  // Line 389
    } else {
        None
    },
    walletdisplayname: apple_pay_data.payment_method.display_name.clone(),
    walletsource: "APPLEPAY".to_string(),  // Line 397

---

3. MEDIUM - Cryptogram empty check reliability

The cryptogram presence check uses is_empty() which may not catch all invalid placeholder values.

crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs

Lines 373-375

    let cryptogram = apple_pay_decrypted_data
        .payment_data
        .online_payment_cryptogram
        .clone();
    let is_cryptogram_3ds = !cryptogram.peek().is_empty();

The test data shows "AAAAAA==" as a valid cryptogram. Ensure this check handles edge cases where the cryptogram might be a placeholder like " " or other invalid values. Consider validating the base64 decoding or using a minimum length check.

---

4. MEDIUM - Magic number for tokenisedpayment flag

The value "1" for tokenisedpayment is hardcoded without context. This should be a named constant.

crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs

Lines 385-390

    tokenisedpayment: if is_cryptogram_3ds {
        Some("1".to_string())  // What does "1" mean?
    } else {
        None
    },

Suggested fix:

    const TRUSTPAYMENTS_TOKENIZED_PAYMENT_ENABLED: &str = "1";
    // Add documentation explaining this is a TrustPayments-specific flag
    
    // Then use:
    tokenisedpayment: if is_cryptogram_3ds {
        Some(TRUSTPAYMENTS_TOKENIZED_PAYMENT_ENABLED.to_string())
    } else {
        None
    },

File: crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs [hunk 2]
```diff
@@ -339,6 +356,68 @@ impl<T: PaymentMethodDataTypes + Debug + Sync + Send + 'static + Serialize>
...
+                            "Trust Payments requires pre-decrypted Apple Pay data; \
+                             encrypted Apple Pay tokens are not supported.",
+                        )
+                    })?;
+
+                let month = apple_pay_decrypted_data.get_expiry_month();
+                let four_digit_year = apple_pay_decrypted_data.get_four_digit_expiry_year();
+                // Trust Payments expects MM/YYYY format
+                let expirydate =
+                    Secret::new(format!("{}/{}", month.peek(), four_digit_year.peek()));
```
Comment (AkshayaFoiger) [inline crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs:381]: NIT: Use predefined functions

File: crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs [hunk 3]
```diff
@@ -339,6 +356,68 @@ impl<T: PaymentMethodDataTypes + Debug + Sync + Send + 'static + Serialize>
...
+                    } else {
+                        None
+                    },
+                    tokentype: if is_cryptogram_3ds {
+                        Some("APPLEPAY".to_string())
+                    } else {
+                        None
+                    },
+                    walletdisplayname: apple_pay_data.payment_method.display_name.clone(),
+                    walletsource: "APPLEPAY".to_string(),
```
Comment (AkshayaFoiger) [inline crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs:418]: NIT: No magic strings, declare enums or if required declare consts and use it .

**Comment by muditbhutani** [comment]
@prism-review-bot please review this pr

File: crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs (hunk 2 above)
Comment (pratikjuspay) [reply crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs]: Done — replaced the inline `format!("{}/{}", month.peek(), four_digit_year.peek())` with a new `get_expiry_date_as_mmyyyy(separator)` helper added to `ApplePayDecryptedData` in `payment_method_data.rs`, consistent with the existing `get_expiry_date_as_mmyy` and `get_expiry_date_as_yyyymm` helpers.

File: crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs (hunk 3 above)
Comment (pratikjuspay) [reply crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs]: Done — defined three named constants inside the Apple Pay match arm: `WALLET_SOURCE = "APPLEPAY"`, `DEFAULT_ECI = "07"`, and `TOKENISED_PAYMENT_ENABLED = "1"`. All three magic strings are now replaced with these constants throughout the arm.

**Comment by pratikjuspay** [comment]
@Shubhodip900! Addressed all four points:

1. **ECI default value** — Replaced `"07"` with a `DEFAULT_ECI` constant defined inside the Apple Pay match arm, with a doc comment explaining it means Apple Pay non-3DS (liability shift via TAVV).

2. **Duplicate `"APPLEPAY"` strings** — Replaced both occurrences with a single `WALLET_SOURCE` constant.

3. **`is_empty()` cryptogram check** — The `online_payment_cryptogram` field on `ApplePayDecryptedData` is `Secret<String>` (not `Option<Secret<String>>`), so it is always present. `is_empty()` is the correct sentinel here — a missing/invalid cryptogram will be an empty string. The `"AAAAAA=="` you mentioned in test data is a valid non-empty cryptogram representing a real (test) TAVV, which correctly passes the check. A more sophisticated base64 validation would be overkill given UCS never generates this value itself; it comes directly from the Apple Pay decryption response.

4. **`tokenisedpayment` flag** — Replaced `"1"` with a `TOKENISED_PAYMENT_ENABLED` constant with a doc comment explaining it is the Trust Payments flag indicating a tokenised/wallet payment.

**Review by AkshayaFoiger — DISMISSED** [review]

**Review by Shubhodip900 — DISMISSED** [review]

File: crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs [hunk 4]
```diff
@@ -193,6 +194,22 @@ pub struct TrustpaymentsGooglePayData {
...
+#[derive(Debug, Serialize)]
+pub struct TrustpaymentsApplePayData {
+    pub pan: Secret<String>,
+    pub expirydate: Secret<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub tavv: Option<Secret<String>>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub eci: Option<String>,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub tokenisedpayment: Option<String>,
```
Comment (AkshayaFoiger) [inline crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs]: From the documentation, it looks like tokenisedpayment and tokentype are response fields. Can we double-check the request fields, or are we referring to a different document?

**Review by AkshayaFoiger — CHANGES_REQUESTED** [review]

File: crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs (hunk 4 above)
Comment (pratikjuspay) [reply crates/integrations/connector-integration/src/connectors/trustpayments/transformers.rs]: @AkshayaFoiger you are right, as per the [Apple Pay direct AUTH integration](https://help.trustpayments.com/hc/en-us/articles/15918529481361-Apple-Pay-direct-AUTH-integration) docs, `tokenisedpayment` and `tokentype` are **response-only** fields (listed only under the AUTH response spec, not the request spec). Removed both from `TrustpaymentsApplePayData` (request struct) and the populate logic, and moved them into `TrustpaymentsAuthResponse` for deserialization.

**Review by AkshayaFoiger — DISMISSED** [review]

**Review by maverox — DISMISSED** [review]

**Review by AkshayaFoiger — APPROVED** [review]

**Review by Shubhodip900 — APPROVED** [review]

**Review by maverox — APPROVED** [review]
